### PR TITLE
Update to use HyperlaneDomain

### PR DESCRIPTION
**Pull request by mattiecnvr**
### Description

Changed out all the places we were using domain id + chain name to just use `HyperlaneDomain`. After talking with @tkporter realized this needed to be a little more complicated than our existing `HyperlaneDomain` in order to correctly support unknown chains. There's a couple of todo spots for this that I have left but I also noticed that we had some sharp edges around this that might have been hard to notice otherwise.

I do intentionally maintain strings/u32 values at the boundaries for both Config types and databases as in both cases that is the underlying value type and the code designed to manage that should be able to handle it.

Ultimately what we should do to fully support custom domains is to take more information in the config if it is not a known one. That way we can have parity with the information we have baked in and what they provide.

Also while this looks like a lot of changes it was pretty quick to do. Mostly just the compiler guiding me as to what needed updating next.

### Drive-by changes

Created a new `HyperlaneDomainImpl` type that can be used to determine what implementation we have.

### Backward compatibility

_Are these changes backward compatible?_

Yes

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

_What kind of testing have these changes undergone?_

Unit Tests

### CODE REVIEW

File: rust/agents/relayer/src/relayer.rs [hunk 1]
```diff
@@ -64,8 +64,10 @@ impl BaseAgent for Relayer {
         let blacklist = parse_matching_list(&settings.blacklist);
         info!(whitelist = %whitelist, blacklist = %blacklist, "Whitelist configuration");
 
+        // TODO: we should support unknown chains
```
Comment (tkporter) [inline rust/agents/relayer/src/relayer.rs:67]: This feels like a big regression that I'm really hesitant to merge with especially with our push for permissionless deployment (see https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1450). Iiuc previously someone could run a relayer or validator for a chain by only making config changes, and now it'd require code changes.

I think an easy way to keep the same functionality could be:
```
let origin_chain = core.settings.chain_setup(&settings.originchainname)?.domain()?;
```
And to make the `domain` fn pub. Same idea for the validator. Wdyt?

Alternatively could consider iterating through `core.mailboxes` or something and finding the `HyperlaneDomain` that way

**Review by tkporter — COMMENTED** [review]
Everything LGTM except for what I mentioned in my comment, I think it's important to preserve the ability to run the relayer / validator without code changes.

I think as it is rn, we wouldn't be able to run the validator or relayer for local Fuel development either (because there is no Fuel-based KnownHyperlaneDomain)

File: rust/agents/relayer/src/relayer.rs (hunk 1 above)
Comment (mattiecnvr) [reply rust/agents/relayer/src/relayer.rs]: Yeah, I like this, done.

**Review by tkporter — APPROVED** [review]
LGTM! great job

**Pull request merged**